A pod with an unsigned image can't be deleted

**Describe the bug**

When update deployment from one image to other, connoisseur validate new image and then validate old image. And if old image doesn't have signature in repo, old ReplicaSet stuck with old running pod.

**Expected behavior**

Pod without signature must be removed without validation.

**Versions:**

 - Kubernetes Cluster: RKE v1.24.8
 - Container registry: Nexus 3.58.1
 - Connaisseur: 3.0.0
 - ArgoCD: v2.6.7

**Additional:**
```
kubectl get replicaset -n test -owide | grep my-app

NAME                     DESIRED   CURRENT   READY   AGE     CONTAINERS            IMAGES
my-app-8646d49c66        1         1         1       8m55s   my-app                my-docker.nexus.int.my-dev.ru/my-app:3.0.5@sha256:...3e45
my-app-ffb5fbf98         1         1         1       60m     my-app                my-docker.nexus.int.my-dev.ru/my-app:3.0.6@sha256:...0703
```
**logs:**

In logs a version update from 3.0.5 to 3.0.6
```
"COSIGN output of trust root 'my-docker' for image'my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703': RETURNCODE: 0; STDOUT: {\"critical\":{\"identity\":{\"docker-reference\":\"my-docker.nexus.int.my-dev.ru/myapp\"},\"image\":{\"docker-manifest-digest\":\"sha256:...703\"},\"type\":\"cosign container image signature\"},\"optional\":null}\n; STDERR: \nVerification for my-docker.nexus.int.my-dev.ru/myapp@sha256:...703 --\nThe following checks were performed on each of these signatures:\n  - The cosign claims were validated\n  - Existence of the claims in the transparency log was verified offline\n  - The signatures were verified against the specified public key\n"
"successful verification of image \"my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703\"", "admission_review": {"user": "u-gwg2v2oh7l", "operation": "UPDATE", "kind": "Deployment", "name": "myapp", "namespace": "test"}, "image": "my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703"
"automatic child approval for \"my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703\".", "admission_review": {"user": "system:serviceaccount:kube-system:deployment-controller", "operation": "CREATE", "kind": "ReplicaSet", "name": "myapp-65bcdfd7bd", "namespace": "test"}, "image": "my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703"
"automatic child approval for \"my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703\".", "admission_review": {"user": "system:serviceaccount:kube-system:deployment-controller", "operation": "UPDATE", "kind": "ReplicaSet", "name": "myapp-86997f749c", "namespace": "test"}, "image": "my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703"
"automatic child approval for \"my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703\".", "admission_review": {"user": "system:serviceaccount:kube-system:deployment-controller", "operation": "UPDATE", "kind": "ReplicaSet", "name": "myapp-65bcdfd7bd", "namespace": "test"}, "image": "my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703"
"automatic child approval for \"my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703\".", "admission_review": {"user": "system:serviceaccount:kube-system:replicaset-controller", "operation": "CREATE", "kind": "Pod", "name": "myapp-65bcdfd7bd-", "namespace": "test"}, "image": "my-docker.nexus.int.my-dev.ru/myapp:3.0.6@sha256:...703"

"COSIGN output of trust root 'my-docker' for image'my-docker.nexus.int.my-dev.ru/myapp:3.0.5@sha256:...e45': RETURNCODE: 12; STDOUT: ; STDERR: Error: no matching signatures:\n\nmain.go:69: error during command execution: no matching signatures:\n"
"{'message': 'No trust data for image \"my-docker.nexus.int.my-dev.ru/myapp:3.0.5@sha256:...e45\".', 'context': {'trust_data_type': 'dev.cosignproject.cosign/signature', 'stderr': 'Error: no matching signatures:\\n\\nmain.go:69: error during command execution: no matching signatures:\\n', 'image': 'my-docker.nexus.int.my-dev.ru/myapp:3.0.5@sha256:...e45', 'trust_root': 'my-docker', 'detection_mode': False}}"
```

--- a/.pylintrc
+++ b/.pylintrc
@@ -76,6 +76,7 @@ disable=raw-checker-failed,
         R1710,
         W0703,
         W0511,
+        W0212,
 
 # Enable the message, report, category or checker with the given id(s). You can
 # either give multiple identifier separated by comma (,) or put this option

--- a/connaisseur/flask_application.py
+++ b/connaisseur/flask_application.py
@@ -173,18 +173,28 @@ async def __validate_image(
     # if automatic_unchanged_approval is enabled, admit resource updates
     # if they do not alter the resource's image reference(s)
     # https://github.com/sse-secure-systems/connaisseur/issues/820
+    # exception: if a resource increases its replicas from 0 to >0, it is
+    # not automatically approved, even if the image reference is unchanged.
+    # this is because the resource may have previously been set to 0 and
+    # also changed its image reference, which would have been skipped (see
+    # below). if the replicas are now increased, the image reference would
+    # not be checked, which is not desired.
     if (
         feature_flag_on(const.AUTOMATIC_UNCHANGED_APPROVAL)
         and admission_request.operation.upper() == "UPDATE"
+        and not (
+            admission_request.wl_object.kind in ["ReplicaSet", "Deployment"]
+            and admission_request.wl_object.replicas > 0
+            and admission_request.old_wl_object.replicas == 0
+        )
+        and image in admission_request.old_wl_object.containers.values()
     ):
-        old_images = admission_request.old_wl_object.containers.values()
-        if image in old_images:
-            logging.info(
-                'automatic approval for unchanged image "%s".',
-                original_image,
-                extra=logging_context,
-            )
-            return
+        logging.info(
+            'automatic approval for unchanged image "%s".',
+            original_image,
+            extra=logging_context,
+        )
+        return
 
     # child resources have mutated image names, as their parents got mutated
     # before their creation. this may result in mismatch of rules or duplicate
@@ -196,6 +206,22 @@ async def __validate_image(
     ):
         logging.info(
             'automatic child approval for "%s".', original_image, extra=logging_context
+        )
+        return
+
+    # if replicas are set to 0, skip verification, since no pods will be created.
+    # otherwise we might run into problems with replicasets having an image
+    # without signature. then reducing the replicas to 0 will never be allowed
+    # by Connaisseur.
+    if (
+        admission_request.operation.upper() == "UPDATE"
+        and admission_request.kind in ["ReplicaSet", "Deployment"]
+        and admission_request.wl_object.replicas == 0
+    ):
+        logging.info(
+            'replicas set to 0. skipping verification of "%s".',
+            original_image,
+            extra=logging_context,
         )
         return
 

--- a/connaisseur/workload_object.py
+++ b/connaisseur/workload_object.py
@@ -22,11 +22,13 @@ class WorkloadObject:
     def __new__(
         cls, request_object: dict, namespace: str
     ):  # pylint: disable=unused-argument
-        if request_object["kind"] == "Pod":
-            return super(WorkloadObject, cls).__new__(Pod)
-        elif request_object["kind"] == "CronJob":
-            return super(WorkloadObject, cls).__new__(CronJob)
-        return super(WorkloadObject, cls).__new__(WorkloadObject)
+        workload_object_class = {
+            "CronJob": CronJob,
+            "Deployment": WorkloadObjectWithReplicas,
+            "Pod": Pod,
+            "ReplicaSet": WorkloadObjectWithReplicas,
+        }.get(request_object["kind"], WorkloadObject)
+        return super(WorkloadObject, cls).__new__(workload_object_class)
 
     def __init__(self, request_object: dict, namespace: str):
         self.kind = request_object["kind"]
@@ -125,3 +127,9 @@ class CronJob(WorkloadObject):
     @property
     def spec(self):
         return self._spec["jobTemplate"]["spec"]["template"]["spec"]
+
+
+class WorkloadObjectWithReplicas(WorkloadObject):
+    @property
+    def replicas(self):
+        return self._spec.get("replicas")

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -352,6 +352,9 @@ def adm_req_samples(m_ad_schema_path):
             "auto_approval",
             "invalid",
             "auto_update_approval",
+            "replicaset_to_zero",
+            "replicaset_from_zero",
+            "replica_from_non_zero",
         )
     ]
 

--- a/tests/data/sample_admission_requests/ad_request_replica_from_non_zero.json
+++ b/tests/data/sample_admission_requests/ad_request_replica_from_non_zero.json
@@ -1,0 +1,182 @@
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "request": {
+        "uid": "8c6c8be6-479d-41f4-82ba-158998cdcbf8",
+        "kind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "resource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "requestKind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "requestResource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "name": "test-deploy-85597768cc",
+        "namespace": "default",
+        "operation": "UPDATE",
+        "userInfo": {
+            "username": "system:serviceaccount:kube-system:deployment-controller",
+            "uid": "3b24cf0d-787d-4f00-88cc-bfebff939a49",
+            "groups": [
+                "system:serviceaccounts",
+                "system:serviceaccounts:kube-system",
+                "system:authenticated"
+            ]
+        },
+        "object": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 2,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "test/deny-image",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "oldObject": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 1,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "test/deny-image",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "dryRun": false,
+        "options": {
+            "kind": "UpdateOptions",
+            "apiVersion": "meta.k8s.io/v1"
+        }
+    }
+}

--- a/tests/data/sample_admission_requests/ad_request_replicaset_from_zero.json
+++ b/tests/data/sample_admission_requests/ad_request_replicaset_from_zero.json
@@ -1,0 +1,182 @@
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "request": {
+        "uid": "8c6c8be6-479d-41f4-82ba-158998cdcbf8",
+        "kind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "resource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "requestKind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "requestResource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "name": "test-deploy-85597768cc",
+        "namespace": "default",
+        "operation": "UPDATE",
+        "userInfo": {
+            "username": "system:serviceaccount:kube-system:deployment-controller",
+            "uid": "3b24cf0d-787d-4f00-88cc-bfebff939a49",
+            "groups": [
+                "system:serviceaccounts",
+                "system:serviceaccounts:kube-system",
+                "system:authenticated"
+            ]
+        },
+        "object": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 2,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "test/deny-image",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "oldObject": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 0,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "test/deny-image",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "dryRun": false,
+        "options": {
+            "kind": "UpdateOptions",
+            "apiVersion": "meta.k8s.io/v1"
+        }
+    }
+}

--- a/tests/data/sample_admission_requests/ad_request_replicaset_to_zero.json
+++ b/tests/data/sample_admission_requests/ad_request_replicaset_to_zero.json
@@ -1,0 +1,182 @@
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "request": {
+        "uid": "8c6c8be6-479d-41f4-82ba-158998cdcbf8",
+        "kind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "resource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "requestKind": {
+            "group": "apps",
+            "version": "v1",
+            "kind": "ReplicaSet"
+        },
+        "requestResource": {
+            "group": "apps",
+            "version": "v1",
+            "resource": "replicasets"
+        },
+        "name": "test-deploy-85597768cc",
+        "namespace": "default",
+        "operation": "UPDATE",
+        "userInfo": {
+            "username": "system:serviceaccount:kube-system:deployment-controller",
+            "uid": "3b24cf0d-787d-4f00-88cc-bfebff939a49",
+            "groups": [
+                "system:serviceaccounts",
+                "system:serviceaccounts:kube-system",
+                "system:authenticated"
+            ]
+        },
+        "object": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 0,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "node:alpine",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "oldObject": {
+            "kind": "ReplicaSet",
+            "apiVersion": "apps/v1",
+            "metadata": {
+                "name": "test-deploy-85597768cc",
+                "namespace": "default",
+                "uid": "4cce84d2-9bef-4836-8f59-966f86f72cef",
+                "resourceVersion": "119523",
+                "generation": 1,
+                "creationTimestamp": "2023-09-29T08:21:21Z",
+                "labels": {
+                    "app": "test-deploy",
+                    "pod-template-hash": "85597768cc"
+                },
+                "annotations": {
+                    "deployment.kubernetes.io/desired-replicas": "1",
+                    "deployment.kubernetes.io/max-replicas": "2",
+                    "deployment.kubernetes.io/revision": "1"
+                },
+                "ownerReferences": []
+            },
+            "spec": {
+                "replicas": 1,
+                "selector": {
+                    "matchLabels": {
+                        "app": "test-deploy",
+                        "pod-template-hash": "85597768cc"
+                    }
+                },
+                "template": {
+                    "metadata": {
+                        "creationTimestamp": null,
+                        "labels": {
+                            "app": "test-deploy",
+                            "pod-template-hash": "85597768cc"
+                        }
+                    },
+                    "spec": {
+                        "containers": [
+                            {
+                                "name": "node",
+                                "image": "node:alpine",
+                                "command": [
+                                    "ash",
+                                    "-c",
+                                    "sleep 300000"
+                                ],
+                                "resources": {},
+                                "terminationMessagePath": "/dev/termination-log",
+                                "terminationMessagePolicy": "File",
+                                "imagePullPolicy": "IfNotPresent"
+                            }
+                        ],
+                        "restartPolicy": "Always",
+                        "terminationGracePeriodSeconds": 30,
+                        "dnsPolicy": "ClusterFirst",
+                        "securityContext": {},
+                        "schedulerName": "default-scheduler"
+                    }
+                }
+            },
+            "status": {
+                "replicas": 1,
+                "fullyLabeledReplicas": 1,
+                "readyReplicas": 1,
+                "availableReplicas": 1,
+                "observedGeneration": 1
+            }
+        },
+        "dryRun": false,
+        "options": {
+            "kind": "UpdateOptions",
+            "apiVersion": "meta.k8s.io/v1"
+        }
+    }
+}

--- a/tests/test_flask_application.py
+++ b/tests/test_flask_application.py
@@ -17,7 +17,11 @@ from . import conftest as fix
 @pytest.fixture(autouse=True)
 def m_config(monkeypatch, sample_nv1):
     def mock_init(self):
-        self.validators = [sample_nv1, StaticValidator("allow", True)]
+        self.validators = [
+            sample_nv1,
+            StaticValidator("allow", True),
+            StaticValidator("deny", False),
+        ]
         self.policy = [
             {
                 "pattern": "*:*",
@@ -56,6 +60,10 @@ def m_config(monkeypatch, sample_nv1):
             {
                 "pattern": "docker.io/securesystemsengineering/alice-image",
                 "validator": "dockerhub",
+            },
+            {
+                "pattern": "docker.io/test/deny-image:*",
+                "validator": "deny",
             },
         ]
 
@@ -134,10 +142,11 @@ def test_readyz():
 
 @pytest.mark.asyncio
 @pytest.mark.parametrize(
-    "index, out, exception",
+    "index, update_approval, out, exception",
     [
         (
             0,
+            True,
             {
                 "apiVersion": "admission.k8s.io/v1",
                 "kind": "AdmissionReview",
@@ -160,6 +169,7 @@ def test_readyz():
         ),
         (
             1,
+            True,
             {
                 "apiVersion": "admission.k8s.io/v1",
                 "kind": "AdmissionReview",
@@ -171,9 +181,10 @@ def test_readyz():
             },
             fix.no_exc(),
         ),
-        (5, {}, pytest.raises(exc.BaseConnaisseurException)),
+        (5, True, {}, pytest.raises(exc.BaseConnaisseurException)),
         (
             6,
+            True,
             {
                 "apiVersion": "admission.k8s.io/v1",
                 "kind": "AdmissionReview",
@@ -187,11 +198,52 @@ def test_readyz():
         ),
         (
             8,
+            True,
             {
                 "apiVersion": "admission.k8s.io/v1",
                 "kind": "AdmissionReview",
                 "response": {
                     "uid": "6418f614-b7f1-4c57-94bf-14fc30b2d791",
+                    "allowed": True,
+                    "status": {"code": 202},
+                },
+            },
+            fix.no_exc(),
+        ),
+        (
+            9,
+            False,
+            {
+                "apiVersion": "admission.k8s.io/v1",
+                "kind": "AdmissionReview",
+                "response": {
+                    "uid": "8c6c8be6-479d-41f4-82ba-158998cdcbf8",
+                    "allowed": True,
+                    "status": {"code": 202},
+                },
+            },
+            fix.no_exc(),
+        ),
+        (
+            10,
+            True,
+            {},
+            pytest.raises(exc.ValidationError),
+        ),
+        (
+            10,
+            False,
+            {},
+            pytest.raises(exc.ValidationError),
+        ),
+        (
+            11,
+            True,
+            {
+                "apiVersion": "admission.k8s.io/v1",
+                "kind": "AdmissionReview",
+                "response": {
+                    "uid": "8c6c8be6-479d-41f4-82ba-158998cdcbf8",
                     "allowed": True,
                     "status": {"code": 202},
                 },
@@ -207,13 +259,13 @@ async def test_admit(
     m_request,
     m_expiry,
     m_trust_data,
+    update_approval,
     out,
     exception,
 ):
     session = aiohttp.ClientSession()
+    monkeypatch.setenv("AUTOMATIC_UNCHANGED_APPROVAL", str(update_approval))
     with exception:
-        if index == 8:
-            monkeypatch.setenv("AUTOMATIC_UNCHANGED_APPROVAL", "true")
         with aioresponses() as aio:
             aio.get(re.compile(r".*"), callback=fix.async_callback, repeat=True)
             response = await pytest.fa.__admit(

